### PR TITLE
Added mechanism for group adding of records

### PR DESCRIPTION
**Pull request by areksolek**
Added mechanism for group adding of records

### CODE REVIEW

File: app/RecordAddsTemplates.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	{
+		return new $className();
+	}
+
+	/**
+	 * List of available templates.
+	 *
+	 * @return array
+	 */
+	public static function getTemplatesList()
```
Comment (mariuszkrzaczkowski) [inline app/RecordAddsTemplates.php]:
```suggestion
	public static function getTemplatesList(): array
```

File: app/RecordAddsTemplates.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * Record adds templates class.
+ */
+class RecordAddsTemplates
+{
+	/**
+	 * Get record adds templates instance.
+	 *
+	 * @param string $className
+	 *
+	 * @return RecordAddsTemplates\Paccar|null
```
Comment (mariuszkrzaczkowski) [inline app/RecordAddsTemplates.php]: ?

File: app/RecordAddsTemplates.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	 *
+	 * @return array
+	 */
+	public static function getTemplatesList()
+	{
+		$listTemplates = [];
+		foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator(ROOT_DIRECTORY . \DIRECTORY_SEPARATOR . 'app' . \DIRECTORY_SEPARATOR . 'RecordAddsTemplates', \RecursiveDirectoryIterator::SKIP_DOTS), \RecursiveIteratorIterator::SELF_FIRST) as $item) {
+			if ($item->isFile()) {
+				$fileName = $item->getBasename('.php');
+				$instance = static::getInstance("\\App\\RecordAddsTemplates\\$fileName");
```
Comment (mariuszkrzaczkowski) [inline app/RecordAddsTemplates.php]:
```suggestion
				$instance = static::getInstance($fileName);
```

File: modules/Vtiger/views/RecordAddsTemplates.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	/** {@inheritdoc} */
+	public $modalSize = 'c-modal-xxl';
+
+	/** Record adds instance @var \App\RecordAddsTemplates\ */
+	private $recordAddsInstance;
+
+	/** {@inheritdoc} */
+	public function checkPermission(App\Request $request)
+	{
+		$this->recordAddsInstance = \App\RecordAddsTemplates::getInstance("\\App\\RecordAddsTemplates\\{$request->getByType('recordAddsType', 'ClassName')}");
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/RecordAddsTemplates.php]:
```suggestion
		$this->recordAddsInstance = \App\RecordAddsTemplates::getInstance($request->getByType('recordAddsType', 'ClassName'));
```

File: modules/Vtiger/views/RecordAddsTemplates.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	public $showFooter = true;
+
+	/** {@inheritdoc} */
+	public $modalSize = 'c-modal-xxl';
+
+	/** Record adds instance @var \App\RecordAddsTemplates\ */
+	private $recordAddsInstance;
+
+	/** {@inheritdoc} */
+	public function checkPermission(App\Request $request)
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/RecordAddsTemplates.php:28]: add \App\Config::main('isActiveRecordTemplate')

File: modules/Vtiger/views/RecordAddsTemplates.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		$viewer->assign('MODE', '');
+		$viewer->assign('RECORD_TEMPLATE', $request->getByType('recordAddsType', 'ClassName'));
+		$viewer->assign('BLOCK_LIST', $this->recordAddsInstance->getBlocks());
+		$viewer->assign('RECORD', null);
+		$viewer->assign('SCRIPTS', $this->getModalScripts($request));
+		$viewer->view('RecordAddsTemplates.tpl');
+	}
+
+	/** {@inheritdoc} */
+	public function getModalScripts(App\Request $request)
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/RecordAddsTemplates.php]: ?

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

File: public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+							app.errorLog(textStatus, errorThrown);
+							aDeferred.resolve(false);
+						});
+				}
+			});
+		},
+
+		/**
+		 * Register events function
+		 * @param modalContainer
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
+/* {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} */
+$.Class(
+	'Base_RecordAddsTemplates_JS',
+	{},
+	{
+		editInstance: false,
+		/**
+		 * Function listener to send a form
+		 * @param container
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js]: ?

File: app/RecordAddsTemplates.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+ * Record adds templates class.
+ */
+class RecordAddsTemplates
+{
+	/**
+	 * Get record adds templates instance.
+	 *
+	 * @param string $className
+	 *
+	 * @return RecordAddsTemplates\Object
```
Comment (mariuszkrzaczkowski) [inline app/RecordAddsTemplates.php]:
```suggestion
	 * @return Object
```

File: layouts/basic/modules/Vtiger/RecordAddsTemplates.tpl [hunk 10]
```diff
@@ -0,0 +1,14 @@
+{*<!-- {[The file is published on the basis of YetiForce Public License 3.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} -->*}
+{strip}
+<!-- tpl-Base-RecordAddsTemplates -->
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/RecordAddsTemplates.tpl]:
```suggestion
<!-- tpl-Base-Modals-RecordAddsTemplates -->
```

File: layouts/basic/modules/Vtiger/BodyHeader.tpl [hunk 11]
```diff
@@ -241,6 +241,48 @@
...
+									href="#"
+									role="button">
+										<span class="mdi mdi-plus-box-multiple" title="{\App\Language::translate('LBL_BATCH_ADDING_RECORDS')}"></span>
+									</a>
+									<div class="dropdown-menu p-0 u-max-w-sm-100 u-min-w-300px" aria-labelledby="recordAddsTemplate" role="list">
+										<div class="container-fluid d-block p-2 u-max-w-xsm-100 px-2">
+											{foreach from=$LIST_TEMPLATES item=TEMPLATE_VALUE}
+												<div class="row">
+													<div class="col-12 u-bg-light-darken">
+														<a class="showModal text-decoration-none u-fs-sm text-secondary  d-block" data-url="index.php?module={$QUALIFIED_MODULE}&view=RecordAddsTemplates&recordAddsType={$TEMPLATE_VALUE['templateName']}"
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/BodyHeader.tpl]: $QUALIFIED_MODULE ?

File: app/RecordAddsTemplates.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	public static function getTemplatesList(): array
+	{
+		$listTemplates = [];
+		$pathDirectory = ROOT_DIRECTORY . \DIRECTORY_SEPARATOR . 'app' . \DIRECTORY_SEPARATOR . 'RecordAddsTemplates';
+		if (file_exists($pathDirectory)) {
+			foreach (new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($pathDirectory, \RecursiveDirectoryIterator::SKIP_DOTS), \RecursiveIteratorIterator::SELF_FIRST) as $item) {
+				if ($item->isFile()) {
+					$fileName = $item->getBasename('.php');
+					$instance = static::getInstance($fileName);
+					$listTemplates[] = ['templateName' => $fileName, 'label' => $instance->label, 'icon' => $instance->icon];
```
Comment (mariuszkrzaczkowski) [inline app/RecordAddsTemplates.php]:
```suggestion
					$listTemplates[] = static::getInstance($fileName);
```

File: modules/Vtiger/actions/RecordAddsTemplates.php [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+use App\Request;
+
+/**
+ * Record adds templates actions class.
+ */
+class Vtiger_RecordAddsTemplates_Action extends \App\Controller\Action
+{
+	/** @var \App\RecordAddsTemplates\Object Record adds instance. */
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/actions/RecordAddsTemplates.php]:
```suggestion
	/** @var Object Record adds instance. */
```

File: layouts/basic/modules/Vtiger/BodyHeader.tpl [hunk 14]
```diff
@@ -241,6 +241,48 @@
...
+													</div>
+												</div>
+											{/foreach}
+										</div>
+									</div>
+								</div>
+							</div>
+						{elseif $LIST_TEMPLATES}
+							<div class="o-action-menu__item">
+								<a class="c-header__btn ml-2 showModal btn-light btn js-popover-tooltip" role="button" data-url="index.php?module={$QUALIFIED_MODULE}&view=RecordAddsTemplates&recordAddsType={$LIST_TEMPLATES[0]['templateName']}"
```
Comment (mariuszkrzaczkowski) [inline layouts/basic/modules/Vtiger/BodyHeader.tpl]: $QUALIFIED_MODULE ?

File: modules/Vtiger/actions/RecordAddsTemplates.php [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+	{
+		if (!\App\Config::main('isActiveRecordTemplate')) {
+			throw new \App\Exceptions\NoPermitted('LBL_PERMISSION_DENIED');
+		}
+		$this->recordAddsInstance = \App\RecordAddsTemplates::getInstance($request->getByType('recordAddsType', 'ClassName'));
+		$this->recordAddsInstance->checkPermission();
+	}
+
+	/** {@inheritdoc} */
+	public function process(Request $request)
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/actions/RecordAddsTemplates.php]:
```suggestion
	public function process(Request $request): void
```

File: modules/Vtiger/views/RecordAddsTemplates.php [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+ */
+class Vtiger_RecordAddsTemplates_View extends \App\Controller\Modal
+{
+	/** {@inheritdoc} */
+	public $showFooter = true;
+
+	/** {@inheritdoc} */
+	public $modalSize = 'c-modal-xxl';
+
+	/** @var \App\RecordAddsTemplates\Object Record adds instance. */
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/RecordAddsTemplates.php]:
```suggestion
	/** @var Object Record adds instance. */
```

File: modules/Vtiger/views/RecordAddsTemplates.php [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+	/** {@inheritdoc} */
+	public function preProcessAjax(App\Request $request)
+	{
+		$this->modalIcon = $this->recordAddsInstance->icon;
+		$this->pageTitle = $this->recordAddsInstance->label;
+		parent::preProcessAjax($request);
+	}
+
+	/** {@inheritdoc} */
+	public function process(App\Request $request)
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/RecordAddsTemplates.php]:
```suggestion
	public function process(App\Request $request):void
```

File: modules/Vtiger/views/RecordAddsTemplates.php [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+		$viewer->assign('MODULE_NAME', $request->getModule());
+		$viewer->assign('RECORD_STRUCTURE', $this->recordAddsInstance->getFields());
+		$viewer->assign('MODULE_FORM', array_keys($this->recordAddsInstance->modulesFieldsMap));
+		$viewer->assign('VIEW', 'recordTemplate');
+		$viewer->assign('MODE', '');
+		$viewer->assign('RECORD_TEMPLATE', $request->getByType('recordAddsType', 'ClassName'));
+		$viewer->assign('BLOCK_LIST', $this->recordAddsInstance->getBlocks());
+		$viewer->assign('RECORD', null);
+		$viewer->assign('SCRIPTS', $this->getModalScripts($request));
+		$viewer->view('RecordAddsTemplates.tpl');
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/RecordAddsTemplates.php]:
```suggestion
		$viewer->view('Modals/RecordAddsTemplates.tpl');
```

File: modules/Vtiger/views/RecordAddsTemplates.php [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+		$viewer = $this->getViewer($request);
+		$viewer->assign('MODULE_NAME', $request->getModule());
+		$viewer->assign('RECORD_STRUCTURE', $this->recordAddsInstance->getFields());
+		$viewer->assign('MODULE_FORM', array_keys($this->recordAddsInstance->modulesFieldsMap));
+		$viewer->assign('VIEW', 'recordTemplate');
+		$viewer->assign('MODE', '');
+		$viewer->assign('RECORD_TEMPLATE', $request->getByType('recordAddsType', 'ClassName'));
+		$viewer->assign('BLOCK_LIST', $this->recordAddsInstance->getBlocks());
+		$viewer->assign('RECORD', null);
+		$viewer->assign('SCRIPTS', $this->getModalScripts($request));
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/RecordAddsTemplates.php]: ?

File: modules/Vtiger/views/RecordAddsTemplates.php [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+		$this->modalIcon = $this->recordAddsInstance->icon;
+		$this->pageTitle = $this->recordAddsInstance->label;
+		parent::preProcessAjax($request);
+	}
+
+	/** {@inheritdoc} */
+	public function process(App\Request $request)
+	{
+		$viewer = $this->getViewer($request);
+		$viewer->assign('MODULE_NAME', $request->getModule());
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/views/RecordAddsTemplates.php]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+			let instanse = this;
+			container.on('click', '.js-modal__save', () => {
+				let params = {};
+				params.action = 'RecordAddsTemplates';
+				let validate = true;
+				container.find('form.js-record-template').each(function () {
+					let form = $(this);
+					instanse.editInstance.registerValidationsFields(form);
+					let formSerializeData = form.serializeFormData();
+					params.module = formSerializeData.module;
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+					params.recordAddsType = formSerializeData.recordAddsType;
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js]: ?

File: public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+		editInstance: false,
+		/**
+		 * Function listener to send a form
+		 * @param {jQuery} container
+		 */
+		registerSubmitForm: function (container) {
+			const aDeferred = $.Deferred();
+			let instanse = this;
+			container.on('click', '.js-modal__save', () => {
+				let params = {};
```
Comment (mariuszkrzaczkowski) [inline public_html/layouts/basic/modules/Vtiger/resources/RecordAddsTemplates.js]:
```suggestion
				let params = {action: 'RecordAddsTemplates', recordAddsType: ...};
```

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

File: modules/Settings/Vtiger/models/Field.php [hunk 24]
```diff
@@ -14,7 +14,6 @@ class Settings_Vtiger_Field_Model extends Vtiger_Field_Model
 	 *
 	 * @var string[]
 	 */
-	public $referenceList = [];
```
Comment (rskrzypczak) [inline modules/Settings/Vtiger/models/Field.php:17]: why you removed this?

**Review by mariuszkrzaczkowski — APPROVED** [review]

**Pull request merged**